pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/BaseClusterIntegrationTest.java

```diff
@@ -16,1402 +16,322 @@
+        .getJSONObject(0)
+        .getLong("value");
   }
 
-  private void checkForInstanceAbsenceFromRoutingTable(final String instanceAddress) {
-    waitForRoutingTablePredicate(new Function<JSONArray, Boolean>() {
+  /**
+   * Wait for all segments for the given table type to become online.
+   *
+   * @param tableType Table type
+   * @param expectedNumSegments Expected number of online segments
+   * @throws Exception
+   */
+  protected void waitForAllSegmentsOnline(@Nonnull CommonConstants.Helix.TableType tableType,
+      final int expectedNumSegments, long timeoutMs)
+      throws Exception {
+    final String tableNameWithType = TableNameBuilder.forType(tableType).tableNameWithType(getTableName());
+    TestUtils.waitForCondition(new Function<Void, Boolean>() {
       @Nullable
       @Override
-      public Boolean apply(@Nullable JSONArray input) {
+      public Boolean apply(@Nullable Void aVoid) {
         try {
-          int tableCount = input.length();
-
-          for (int i = 0; i < tableCount; i++) {
-            JSONObject tableRouting = input.getJSONObject(i);
-            String tableName = tableRouting.getString("tableName");
-            if (tableName.startsWith(getTableName())) {
-              JSONArray routingTableEntries = tableRouting.getJSONArray("routingTableEntries");
-              int routingTableEntryCount = routingTableEntries.length();
-              for (int j = 0; j < routingTableEntryCount; j++) {
-                JSONObject routingTableEntry = routingTableEntries.getJSONObject(j);
-                Iterator<String> routingTableEntryInstances = routingTableEntry.keys();
-
-                // The randomly selected instance should not be in the routing table
-                while (routingTableEntryInstances.hasNext()) {
-                  String instance = routingTableEntryInstances.next();
-                  if (instance.equals(instanceAddress)) {
-                    return false;
-                  }
-                }
+          ExternalView externalView = _helixAdmin.getResourceExternalView(_clusterName, tableNameWithType);
+          Set<String> partitionSet = externalView.getPartitionSet();
+          if (partitionSet.size() == expectedNumSegments) {
+            int numOnlinePartitions = 0;
+            for (String partitionId : partitionSet) {
+              Map<String, String> partitionStateMap = externalView.getStateMap(partitionId);
+              if (partitionStateMap.containsValue("ONLINE")) {
+                numOnlinePartitions++;
               }
             }
+            if (numOnlinePartitions == expectedNumSegments) {
+              return true;
+            }
           }
-
-          return true;
-        } catch (JSONException e) {
-          LOGGER.warn("Caught exception while reading the routing table, will retry", e);
           return false;
+        } catch (Exception e) {
+          return null;
```

Comment:
Depends. Throw exception means we don't want to try again, while return null and false are the same. I kept the old behavior here.